server/src/main/kotlin/core/gameflow/Player.kt

```diff
@@ -1,45 +1,39 @@
 package core.gameflow
 
 import core.Card
+import core.action.bettinground.*
 
+/**
+ * chipsInPot - the amount of player's chips that already went into the pot in previous rbetting rounds,
+ *      not including the chips in player's current bet
+ */
 data class Player(
     val position: Int,
     val stack: Int,
     val holeCards: List<Card> = emptyList(),
+    val bet: Int = 0,
     val chipsInPot: Int = 0,
-    val currentBet: Int = 0,
-    val folded: Boolean = false,
-    val id: Int = position
+    val lastAction: ActionType? = null
 ) {
-    fun withCards(newCards: List<Card>): Player {
-        return this.copy(holeCards = newCards)
+    init {
+        assert(stack >= 0)
+        assert(bet >= 0)
+        assert(chipsInPot >= 0)
     }
 
-    fun afterFold(): Player {
-        return this.copy(folded = true)
-    }
-
-    fun afterAllIn(): Player {
-        return this.copy(currentBet = this.currentBet + this.stack, stack = 0)
-    }
+    val isInGame: Boolean = lastAction != ActionType.FOLD
+    val isAllIn: Boolean = (stack == 0) and isInGame
```

Comment:
maybe we could add a chipsInPot > 0 condition?(or bet > 0 if the current round hasn't ended yet)

Reply:
or previousAction == AllIn? wtf man

Reply:
1) Why would it be impossible for a player to have no bet?

2) What happens when a player makes a call that puts him all in, or posts the blind pre-flop and doesn't have enough chips? I'm not shure whether we should convert his last action to AllIn or leave it as it actually is and simply put their whole stack into bet, that's why i chose to do it this way.

Reply:
ad. 1. In the round that player goes AllIn(for any reason) his bet must be > 0. In the next rounds his chipsInPot must be > 0. 
ad 2. I agree that lastAction whould also be null when the player went all in in the previous round, which makes point 1. even more relevant.

Reply:
Ok, now I see your point, I completely misunderstood your intentions. We can add this condition, however a scenario in which a player starts a hand with empty stack is impossible in the first place (and this is the only case when we have a player with 0 stack, 0 bet and 0 chips in pot).